File: C#/DeepestLeavesSum.cs
//Ian Hays
//01-16-2021
//https://leetcode.com/problems/deepest-leaves-sum/
//SC: O(N) TC: O(N)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int DeepestLeavesSum(TreeNode root) 
    {
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        int result = 0;
        while(queue.Count != 0)
        {
            int m = queue.Count;
            result = 0;
            for(int i = 0; i < m; i++)
            {
                TreeNode node = queue.Dequeue();
                if(node.left != null) queue.Enqueue(node.left);
                if(node.right != null) queue.Enqueue(node.right);
                result += node.val;
            }
        }
        return result;
    }
}

File: C++/DeepestLeavesSum.cpp
//Ian Hays
//01-16-2021
//https://leetcode.com/problems/deepest-leaves-sum/
//SC: O(N) TC: O(N)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int res;
        while(!q.empty()){
            int m = q.size();
            res = 0;
            for(int i = 0; i < m; i++){
                TreeNode* node = q.front();
                q.pop();
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
                res += node->val;
            }
        }
        return res;
    }
};
